crates/valence_command/src/arg_parser.rs

```diff
@@ -0,0 +1,970 @@
+use std::ops::Add;
+
+use thiserror::Error;
+use valence_server::protocol::packets::play::command_tree_s2c::{Parser, StringArg};
+
+pub trait CommandArgSet {
+    fn from_args(args: Vec<String>) -> Self;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ArgLen {
+    Infinite, // can only be the last argument
+    Exact(u32),
+    Within(char), // Example "man this is cool" would be 4 args without this distinction
+    WithinExplicit(char, char), // [man this is cooler] would be 4 arg without this distinction
+}
+
+pub trait CommandArg: Default {
+    type Result: Default;
```

Comment:
This is a bit suspicious to me. What's the reason we're not just assuming `Result = Self` and using newtype wrappers where appropriate? I imagine that would be important for the derive macro.

Reply:
Many of the arg parses simply use strings. For example their are 3 variants of strings: single word, quoted word and greedy string.